File: BOJ_1654/moeun.java
package com.example.demo;

import java.util.Scanner;

public class Main {
    static int n, k;
    static long lt; // 기본형은 객체가 아닌 변수. 스택메모리에 직접 저장
    static long[] arr = new long[10000]; //배열은 참조형. 힙메모리에 인스턴스? 생성. 힙 메모리에 객체를 생성하고, 그 주소를 참조 변수에 저장

    public static void main(String[] args) {
        // Scanner 객체를 선언하므로 new 사용
        Scanner sc = new Scanner(System.in); // System.in - 자바 표준 입력 인스턴스

        k = sc.nextInt();
        n = sc.nextInt();

        lt = 0;

        for (int i = 0; i < k; i++) {
            long a = sc.nextLong();
            arr[i] = a;

            if (a > lt) {
                lt = a;
            }
        }

        if (lengthCheck(lt))
            System.out.println(lt);
        else
            System.out.println(findLength(1, lt));

        sc.close(); // 다쓰고나면 해제해서 메모리 누수 방지
    }

    static boolean lengthCheck(long l) {
        long cou = 0;
        for (int i = 0; i < k; i++) {
            cou += arr[i] / l;
        }
        return cou >= n;
    }

    static long findLength(long m, long l) {
        while (m < l) {
            long mid = (m + l) / 2;
            if (lengthCheck(mid)) {
                m = mid;
            } else {
                l = mid;
            }

            if (l == m + 1) {
                if (lengthCheck(l)) {
                    m = l;
                    break;
                } else {
                    l = m;
                    break;
                }
            }
        }
        return l;
    }
}

File: BOJ_1654/moeun.cpp
#include<iostream>

using namespace std;

int n, k;
long long lt;
long long arr[10000];

bool length_check(long long l) {
	int cou = 0;
	for (int i = 0; i < k; i++) {
		cou += arr[i] / l;
	}
	//cout << l << " " << cou << "\n";

	if (cou >= n)
		return true;
	return false;
}

long long find_length(long long m, long long l) {
	while (m < l) {
		long long mid = (m + l) / 2;
		if (length_check(mid)) {
			m = mid;
		}
		else {
			l = mid;
		}

		if (l == m + 1) { // m보다 l이 1더 클떈 무한루프 돌수있음
			if (length_check(l)) {
				m = l;
				break;
			}
			else {
				l = m;
				break;
			}
		}
	}

	return l;
}

int main() {
	ios::sync_with_stdio(false);
	cin.tie(0);
	
	cin >> k >> n;

	lt = 0;

	for (int i = 0; i < k; i++) {
		long long a;
		cin >> a;
		arr[i] = a;

		if (a > lt) { // 최대길이부터 탐색시작
			lt = a;
		}
	}


	if (length_check(lt)) //최대길이로 바로 적용될때
		cout << lt;
	else
		cout << find_length(1, lt); //안되면 이분탐색 시작

	return 0;
}
